Public/Src/Utilities/Collections/ConcurrentBoundedSortedCollection.cs

```diff
@@ -57,62 +55,49 @@ public ConcurrentBoundedSortedCollection(int capacity, IComparer<TSort> comparer
+            if (m_list.Count == 0 || m_list.Count < Capacity || sort.CompareTo(m_currentMinimum) > 0)
             {
-                // Skip the (hopefully) common case
-                if (m_list.Count == 0 || m_list.Count < Capacity || sort.CompareTo(m_currentMinimum) > 0)
+                lock (m_lock)
                 {
-                    var lockCookie = m_rwl.UpgradeToWriterLock(Timeout.Infinite);
-
-                    try
+                    // Double check locking pattern
+                    if (m_list.Count == 0 || m_list.Count < Capacity || sort.CompareTo(m_currentMinimum) > 0)
                     {
-                        // This check is not required because the SortedList would push low values out when trimming.
-                        // However, omitting this check leads to lock thrashing.
-                        if (m_list.Count == 0 || m_list.Count < Capacity || sort.CompareTo(m_currentMinimum) > 0)
+                        // Avoid overlap
+                        if (m_list.ContainsKey(sort))
                         {
-                            // Avoid overlap
-                            if (m_list.ContainsKey(sort))
-                            {
-                                return false;
-                            }
-
-                            // Insert the new value
-                            m_list.Add(sort, value);
+                            return false;
+                        }
 
-                            // Trim excess
-                            if (m_list.Count > Capacity)
-                            {
-                                m_list.RemoveAt(0);
-                            }
+                        // Insert the new value
+                        m_list.Add(sort, value);
 
-                            if (m_list.Count != 0)
-                            {
-                                m_currentMinimum = m_list.ElementAt(0).Key;
-                            }
+                        // Trim excess
+                        if (m_list.Count > Capacity)
+                        {
+                            m_list.RemoveAt(0);
+                        }
 
-                            return true;
+                        if (m_list.Count != 0)
+                        {
+                            m_currentMinimum = m_list.ElementAt(0).Key;
                         }
-                    }
-                    finally
-                    {
-                        m_rwl.DowngradeFromWriterLock(ref lockCookie);
+
+                        return true;
                     }
                 }
-
-                return false;
-            }
-            finally
-            {
-                m_rwl.ReleaseReaderLock();
             }
+
+            return false;
         }
 
         /// <inheritdoc />
         public IEnumerator<KeyValuePair<TSort, TValue>> GetEnumerator()
         {
-            return m_list.GetEnumerator();
+            lock (m_lock)

```

Comment:
I'm not sure this is actually helpful. Yep, we obtain enumerator under the lock, but we will use it without it causing the race condition.

Reply:
But we clone the list first, so the enumerator will iterate over the clone